### ISSUE TITLE
UuidV7 sequel generation

### ISSUE THREAD

**Issue by delacry** [ramsey/uuid#467]
Hi, I would like to suggest a feature, I want my ID sorted in database by creation time, `symfony/uid` has ULID implementation that has extra feature for this by default, if you generate more than 1 UUID in same millisecond, following UUIDs will be `first_uuid + n` where `n` is how many UUIDs were generated in the same millisecond, that means random portion is random only in the first UUID in that millisecond. For example 3 generated ULIDs in same millisecond:
```
0183b476-0e1e-eec5-6a79-aaff0a071a4f
0183b476-0e1e-eec5-6a79-aaff0a071a50
0183b476-0e1e-eec5-6a79-aaff0a071a51
```

## Background/problem
Problems occurs when you use UUIDv7 as ID and you want to sort records in database by creation time and when you create 2 rows at the same time in for-each loop. For instance; it may be some data import from external source and it would be expected that entities will be sorted just same as they were taken from CSV/XML. Or for example in ecommerce, when you change order status before redirection to payment [created status => waiting for payment status] and since it would be possibly same millisecond, last part of UUID are random bits, so position will be random and no one could tell which entity (OrderStatusHistory) was created first. 

Saving another field with datetime or milliseconds won't help, it would need to be some sub-millisecond precision.

## Proposal/solution
Solution is simple, and I prefer not to have it like this by default, since it may surprise programmers later (as it happened to me) that last part is not always random and it's less secure against guessing the next UUID.

Only change would be to add boolean parameter (I'm not sure about naming the parameter) `Uuid::uuid7(sequel: true/false)`, since first parameter is `dateTime` (default null) and can be omitted, default value for sequel would be false, so `Uuid::uuid7()` will work same, but `Uuid::uuid7(sequel: true)` will look at latest generated UUIDv7, and if millisecond is equal to current millisecond or that from `$dateTime`, it will just return previous UUID + 1;

To avoid BC break in uuid factory interface, `uuid7()` method may stay untouched, and some `uuid7Sequel()` could be introduced.

**Comment by ramsey** [comment]
Thanks for the suggestion! This sounds interesting. I'll give it some thought. :-)

**Comment by delacry** [comment]
You are welcome, I saw it on twitter that you are trying to figure it out how to implement it, so I created the issue since I'm also facing this, I'm considering switching from ULID to UUIDv7, because it's more standardized and I prefer UUID string syntax, now I'm creating new UUID like this `Uuid::fromBytes((new Ulid())->toBinary())` but I don't like it very much, since UUID recognizes every ULID as another UUID version.

Another thing I thought about today is that there may be little inconsistency, I tested it on ULIDs and it works like I imagined, because otherwise it would consume too much RAM, it saves only last UUID, but that's not a big deal, that problem with for-each in import or that example from ecommerce aren't affected, and I don't see really a use-case where this may be a problem.

```php
$january = new DateTimeImmutable('2022-01-01');
$february = new DateTimeImmutable('2022-02-01');

Ulid::generate($january); // 017e12b8-ad80-1496-f602-07a2a64f4721
Ulid::generate($january); // 017e12b8-ad80-1496-f602-07a2a64f4722

Ulid::generate($february); // 017eb25d-d180-45fc-2c6a-55e0b9f68639
Ulid::generate($february); // 017eb25d-d180-45fc-2c6a-55e0b9f6863a

Ulid::generate($january); // 017e12b8-ad80-71aa-7ef5-4e4840f79114
Ulid::generate($january); // 017e12b8-ad80-71aa-7ef5-4e4840f79115
```

**Comment by delacry** [comment]
Hi @ramsey, I did some searching and I figured that there is the 12-bit monotonic sequence for that reason. I just saw great blog post about UUIDv7, https://blog.daveallie.com/ulid-primary-keys#uuidv7 where it is described how it's used.

![image](https://user-images.githubusercontent.com/45132928/194844470-7b3da8fe-10e5-4d65-852f-725207c44f2e.png)

![image](https://user-images.githubusercontent.com/45132928/194844408-2da73e27-95af-4cf5-bdf3-f37e1ac6ac28.png)

In specification it's defined as `subsec_b` and it's also recommended to use it for that reason https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-01.html#name-uuidv7-clock-sequence-usage, but currently that 12 bits are dedicated to randomness I guess? Because in `UuidFactory::uuidFromBytesAndVersion` the `$bytes` is 48 bits time and 80 bits random string, and method `substr_replace($bytes, $clockSeqHiAndReserved, 8, 2)` did not change $bytes value. Only line above will replace version, so that's 80-4=76 bits of randomness currently, so if clock seq will be used as mentioned in specification, 76-12=64 random bits, because `variant` field is probably also dedicated to randomness.

https://github.com/ramsey/uuid/blob/4.x/src/UuidFactory.php#L476

I checked some other implementation, python - https://pypi.org/project/uuid7/ - there are different amount of bits but it also uses clock seq. Then go - https://pkg.go.dev/github.com/uuid6/uuid6go-proto#UUIDv7 - there is `CounterPrecisionLength` in constructor that says how many bits are dedicated to that sequence when multiple UUID are generated at the same time.

The huge disadvantage that specification brought is that it's very flexible and everyone can use it differently, if you allowed modification of every variable in generation of uuid7, problem could occur in reconstruction of uuid, because we don't know how long was the timestamp, in their specs there are Figure 3-5 where `subsec_b` is sometimes used for clock seq and sometimes for microsecond/nanosecond precision, so in reconstruction it would be needed to specify that, what can be problematic in ORM.

That said, I would prefer `subsec_b` to be for monotonic sequence by default, maybe with some boolean parameter that says if it should be omitted and used randomness instead.

**Comment by ramsey** [comment]
You're referring to an older version of the Peabody draft. The newer version is here and goes into more detail on monotonicity. https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#name-monotonicity-and-counters

**Comment by ramsey** [comment]
I've just published version 4.6.0, which ensures monotonicity of version 7 UUIDs, using the same algorithm Symfony uses. Take a look and let me know if this resolves this issue.

Thanks!

https://github.com/ramsey/uuid/releases/tag/4.6.0

**Comment by delacry** [comment]
Great job, thank you! I'm very happy with the solution, since it's relatively safe against brute force guessing/scanning, since it would take ~8 000 000 requests to find next id if it was generated in the same ms, and we can still keep the same amount of random bits (74 I guess). I have tested it and it works like a charm.
